client/src/telethon/_impl/mtsender/sender.py

```diff
@@ -239,58 +235,34 @@ async def _step_until_receive(self, rx: Future[bytes]) -> bytes:
                 return rx.result()
 
     async def step(self) -> None:
-        ticket_number = self._step_counter
-
-        async with self.lock:
-            if self._step_counter == ticket_number:
-                # We're the one to drive IO.
-                self._step_counter += 1
-                await self._step()
-            # else: different task drove IO.
+        if not self._writing:
+            self._writing = True
+            await self._try_fill_write()
+            self._writing = False
+
+        if not self._reading:
+            self._reading = True
+            self._response_event.clear()
+            await self._try_read()
+            self._reading = False
+        else:
+            await self._response_event.wait()
 
     def pop_updates(self) -> list[Updates]:
         updates = self._updates[:]
         self._updates.clear()
         return updates
 
-    async def _step(self) -> None:
-        self._try_fill_write()
-
-        recv_req = asyncio.create_task(self._request_event.wait())
-        recv_data = asyncio.create_task(self._reader.read(MAXIMUM_DATA))
-        send_data = asyncio.create_task(self._do_send())
-        done, pending = await asyncio.wait(
-            (recv_req, recv_data, send_data),
-            timeout=self._next_ping - asyncio.get_running_loop().time(),
-            return_when=FIRST_COMPLETED,
-        )
-
-        if pending:
-            for task in pending:
-                task.cancel()
-            await asyncio.wait(pending)
-
-        if recv_req in done:
-            self._request_event.clear()
-        if recv_data in done:
-            self._on_net_read(recv_data.result())
-        if send_data in done:
-            self._on_net_write()
-        if not done:
+    async def _try_read(self) -> None:
+        try:
+            async with asyncio.timeout(PING_DELAY):
+                recv_data = await self._reader.read(MAXIMUM_DATA)
```

Comment:
This has a similar problem as discussed previously.

If there's reads happening every time before the timeout, pings will never occur. And even if timeouts are sent, they're not always spaced 60s apart.

This is also blocking new requests from being sent until Telegram responds with something!

Reply:
> This is also blocking new requests from being sent until Telegram responds with something!
I did not put a lock on the step, so if the request is ready to be sent, it will be sent immediately